File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audiencelist/AudienceListView.java
package com.trtc.uikit.livekit.component.audiencelist;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Point;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tencent.cloud.tuikit.engine.extension.TUILiveListManager;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomEngine;
import com.tencent.qcloud.tuicore.util.ScreenUtil;
import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.common.DataReporter;
import com.trtc.uikit.livekit.common.LiveKitLogger;
import com.trtc.uikit.livekit.component.audiencelist.service.AudienceListObserver;
import com.trtc.uikit.livekit.component.audiencelist.service.AudienceListService;
import com.trtc.uikit.livekit.component.audiencelist.store.AudienceListState;
import com.trtc.uikit.livekit.component.audiencelist.view.AudienceListPopupDialog;
import com.trtc.uikit.livekit.component.audiencelist.view.adapter.AudienceListIconAdapter;

import java.util.Locale;
import java.util.Set;

@SuppressLint("ViewConstructor")
public class AudienceListView extends FrameLayout {
    private static final int MAX_SHOW_AVATAR_COUNT                               = 2;
    private static final int LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIENCE_LIST  = 190010;
    private static final int LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIENCE_LIST = 191009;

    private final Context                               mContext;
    private       RecyclerView                          mRecycleAudienceList;
    private       AudienceListIconAdapter               mAdapter;
    private       TextView                              mTextAudienceCount;
    private       LinearLayout                          mLayoutAudienceCount;
    private       AudienceListObserver                  mAudienceObserver;
    private final AudienceListService                   mAudienceListService
            = new AudienceListService();
    private final AudienceListState                     mAudienceListState
            = mAudienceListService.mAudienceListState;
    private final Observer<Set<TUIRoomDefine.UserInfo>> mAudienceListObserver
            = this::onAudienceListChange;
    private final Observer<Integer>                     mAudienceCountObserver
            = this::onAudienceCountChange;
    private final Observer<Boolean>                     mRoomDismissedObserver
            = this::onRoomDismissed;
    private       TUIRoomEngine                         mRoomEngine;
    private       AudienceListPopupDialog               mAudienceListPopupDialog;
    private       OnUserItemClickListener               mOnUserItemClickListener;

    private static final LiveKitLogger LOGGER = LiveKitLogger.getComponentLogger("AudienceListView");

    public AudienceListView(Context context) {
        this(context, null);
    }

    public AudienceListView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AudienceListView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        LayoutInflater.from(mContext).inflate(R.layout.audience_list_layout_icon, this, true);
    }

    @Deprecated
    public void init(TUIRoomDefine.RoomInfo roomInfo) {
        init(convertToLiveInfo(roomInfo));
    }

    public void init(TUILiveListManager.LiveInfo liveInfo) {
        mAudienceListService.initRoomInfo(liveInfo);
        reportData(liveInfo.roomId);
    }

    public void setOnUserItemClickListener(OnUserItemClickListener listener) {
        mOnUserItemClickListener = listener;
        if (mAudienceListPopupDialog != null) {
            mAudienceListPopupDialog.setOnUserItemClickListener(listener);
        }
    }

    public void setScreenOrientation(boolean isPortrait) {
        if (mLayoutAudienceCount != null) {
            mLayoutAudienceCount.setEnabled(isPortrait);
        }
        if (mRecycleAudienceList != null) {
            mRecycleAudienceList.setEnabled(isPortrait);
        }
    }

    protected void initView() {
        bindViewId();

        initAudienceCountView();
        initAudienceAvatarView();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        initView();
        addObserver();
        mRoomEngine = TUIRoomEngine.sharedInstance();
        mAudienceObserver = new AudienceListObserver(mAudienceListService);
        mRoomEngine.addObserver(mAudienceObserver);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        removeObserver();
        mRoomEngine.removeObserver(mAudienceObserver);
    }

    private TUILiveListManager.LiveInfo convertToLiveInfo(TUIRoomDefine.RoomInfo roomInfo) {
        TUILiveListManager.LiveInfo liveInfo = new TUILiveListManager.LiveInfo();
        liveInfo.roomId = roomInfo.roomId;
        liveInfo.name = roomInfo.name;
        liveInfo.ownerId = roomInfo.ownerId;
        liveInfo.ownerName = roomInfo.ownerName;
        liveInfo.ownerAvatarUrl = roomInfo.ownerAvatarUrl;
        return liveInfo;
    }

    private void initAudienceCountView() {
        mLayoutAudienceCount.setOnClickListener(view -> showAudienceListPanelView());
    }

    private void bindViewId() {
        mLayoutAudienceCount = findViewById(R.id.ll_audience_count);
        mTextAudienceCount = findViewById(R.id.tv_audience_count);
        mRecycleAudienceList = findViewById(R.id.rv_audience_list);
    }

    protected void addObserver() {
        mAudienceListState.audienceCount.observeForever(mAudienceCountObserver);
        mAudienceListState.audienceList.observeForever(mAudienceListObserver);
        mAudienceListState.isRoomDismissed.observeForever(mRoomDismissedObserver);
    }

    protected void removeObserver() {
        mAudienceListState.audienceCount.removeObserver(mAudienceCountObserver);
        mAudienceListState.audienceList.removeObserver(mAudienceListObserver);
        mAudienceListState.isRoomDismissed.removeObserver(mRoomDismissedObserver);
    }

    @SuppressLint("ClickableViewAccessibility")
    private void initAudienceAvatarView() {
        mRecycleAudienceList.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL,
                false));
        mAdapter = new AudienceListIconAdapter(mContext, mAudienceListState);
        mRecycleAudienceList.setAdapter(mAdapter);
        mRecycleAudienceList.setOnTouchListener(new OnTouchListener() {
            private final Point   point  = new Point();
            private       boolean scroll = false;

            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        point.set((int) event.getX(), (int) event.getY());
                        scroll = false;
                        break;
                    case MotionEvent.ACTION_MOVE:
                        if (Math.abs(event.getX() - point.x) > 10 && !scroll) {
                            scroll = true;
                        }
                        break;
                    case MotionEvent.ACTION_UP:
                        if (!scroll) {
                            showAudienceListPanelView();
                        }
                        break;
                    default:
                        break;
                }
                return false;
            }
        });
    }

    private void showAudienceListPanelView() {
        mAudienceListService.getAudienceList();
        if (mAudienceListPopupDialog == null) {
            mAudienceListPopupDialog = new AudienceListPopupDialog(mContext, mAudienceListState);
            mAudienceListPopupDialog.setOnUserItemClickListener(mOnUserItemClickListener);
        }
        mAudienceListPopupDialog.show();
    }

    private void onRoomDismissed(Boolean isDismissed) {
        if (Boolean.TRUE.equals(isDismissed)) {
            if (mAudienceListPopupDialog != null) {
                mAudienceListPopupDialog.dismiss();
            }
        }
    }

    private void onAudienceCountChange(int userCount) {
        updateAudienceCount();
    }

    private void onAudienceListChange(Set<TUIRoomDefine.UserInfo> userInfo) {
        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) mRecycleAudienceList.getLayoutParams();
        if (userInfo.size() <= MAX_SHOW_AVATAR_COUNT) {
            params.width = RelativeLayout.LayoutParams.WRAP_CONTENT;
        } else {
            params.width = ScreenUtil.dip2px(56);
        }
        mAdapter.updateData();
        updateAudienceCount();
    }

    private void updateAudienceCount() {
        int userCount = mAudienceListState.audienceCount.getValue();
        int listSize = mAudienceListState.audienceList.getValue().size();
        LOGGER.info("updateAudienceCount userCount:" + userCount + " listSize:" + listSize);
        if (userCount >= AudienceListState.ROOM_MAX_SHOW_USER_COUNT) {
            setUserCount(userCount);
        } else {
            setUserCount(listSize);
        }
    }

    @SuppressLint("SetTextI18n")
    private void setUserCount(int count) {
        String text = formatUserCount(count);
        mTextAudienceCount.setText(text);
    }

    private String formatUserCount(int count) {
        String text = "" + count;
        if (count >= 10000) {
            boolean isChinese = Locale.CHINESE.getLanguage().equals(Locale.getDefault().getLanguage());
            if (isChinese) {
                text = String.format(Locale.getDefault(), "%.1f", count / 10000.0F);
                if (text.endsWith(".0")) {
                    text = text.replace(".0", "");
                }
                text = text + "万";
            } else {
                text = String.format(Locale.getDefault(), "%dk", count / 1000);
            }
        }
        return text;
    }

    private void reportData(String roomId) {
        boolean isVoiceRoom = !TextUtils.isEmpty(roomId) && roomId.startsWith("voice_");
        if (isVoiceRoom) {
            DataReporter.reportEventData(LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIENCE_LIST);
        } else {
            DataReporter.reportEventData(LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIENCE_LIST);
        }
    }

    public interface OnUserItemClickListener {
        void onUserItemClick(TUIRoomDefine.UserInfo userInfo);
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audiencelist/AudienceListView.kt
package com.trtc.uikit.livekit.component.audiencelist

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Point
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tencent.cloud.tuikit.engine.extension.TUILiveListManager
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine
import com.tencent.qcloud.tuicore.util.ScreenUtil
import com.trtc.uikit.livekit.R
import com.trtc.uikit.livekit.common.LiveKitLogger
import com.trtc.uikit.livekit.common.reportEventData
import com.trtc.uikit.livekit.common.ui.BasicView
import com.trtc.uikit.livekit.component.audiencelist.view.AudienceListPopupDialog
import com.trtc.uikit.livekit.component.audiencelist.view.adapter.AudienceListIconAdapter
import com.trtc.uikit.livekit.component.audiencelist.viewmodel.AudienceListConstants.LIVEKIT_AUDIENCE_LIST_MAX_SHOW_AVATAR_COUNT
import com.trtc.uikit.livekit.component.audiencelist.viewmodel.AudienceListConstants.LIVEKIT_AUDIENCE_LIST_ROOM_MAX_SHOW_USER_COUNT
import com.trtc.uikit.livekit.component.audiencelist.viewmodel.AudienceListConstants.LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIENCE_LIST
import com.trtc.uikit.livekit.component.audiencelist.viewmodel.AudienceListConstants.LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIENCE_LIST
import io.trtc.tuikit.atomicxcore.api.LiveAudienceStore
import io.trtc.tuikit.atomicxcore.api.LiveEndedReason
import io.trtc.tuikit.atomicxcore.api.LiveListListener
import io.trtc.tuikit.atomicxcore.api.LiveListStore
import io.trtc.tuikit.atomicxcore.api.LiveUserInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.Locale

@SuppressLint("ViewConstructor")
class AudienceListView @JvmOverloads constructor(
    private val context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : BasicView(context, attrs, defStyleAttr) {

    private var textAudienceCount: TextView
    private var layoutAudienceCount: LinearLayout
    private var recycleAudienceList: RecyclerView
    private var adapter: AudienceListIconAdapter? = null

    private var liveListStore: LiveListStore? = null
    private var audienceStore: LiveAudienceStore? = null

    private var audienceListPopupDialog: AudienceListPopupDialog? = null
    private var onUserItemClickListener: OnUserItemClickListener? = null

    private val logger = LiveKitLogger.getComponentLogger("AudienceListView")
    private val liveListListener = object : LiveListListener() {
        override fun onLiveEnded(liveId: String, reason: LiveEndedReason, message: String) {
            if (liveId != roomId) return
            audienceListPopupDialog?.dismiss()
        }
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.audience_list_layout_icon, this, true)
        layoutAudienceCount = findViewById(R.id.ll_audience_count)
        textAudienceCount = findViewById(R.id.tv_audience_count)
        recycleAudienceList = findViewById(R.id.rv_audience_list)
    }

    @Deprecated("Use init(TUILiveListManager.LiveInfo) instead")
    fun init(roomInfo: TUIRoomDefine.RoomInfo) {
        init(convertToLiveInfo(roomInfo))
    }

    fun init(liveInfo: TUILiveListManager.LiveInfo) {
        super.init(liveInfo.roomId)
        initView()
        reportData(liveInfo.roomId)
    }

    fun setOnUserItemClickListener(listener: OnUserItemClickListener?) {
        onUserItemClickListener = listener
        audienceListPopupDialog?.setOnUserItemClickListener(listener)
    }

    fun setScreenOrientation(isPortrait: Boolean) {
        layoutAudienceCount.isEnabled = isPortrait
        recycleAudienceList.isEnabled = isPortrait
    }

    override fun initStore() {
        liveListStore = LiveListStore.shared()
        audienceStore = LiveAudienceStore.create(liveID = roomId)
    }

    override fun addObserver() {
        liveListStore?.addLiveListListener(liveListListener)
        subscribeStateJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                audienceStore?.liveAudienceState?.audienceList?.collect { audienceList ->
                    onAudienceListChange(audienceList)
                }
            }
            launch {
                audienceStore?.liveAudienceState?.audienceCount?.collect { audienceCount ->
                    updateAudienceCount()
                }
            }
        }
    }

    override fun removeObserver() {
        liveListStore?.removeLiveListListener(liveListListener)
        subscribeStateJob?.cancel()
    }

    private fun initView() {
        initAudienceCountView()
        initAudienceAvatarView()
    }

    private fun convertToLiveInfo(roomInfo: TUIRoomDefine.RoomInfo): TUILiveListManager.LiveInfo {
        return TUILiveListManager.LiveInfo().apply {
            roomId = roomInfo.roomId
            name = roomInfo.name
            ownerId = roomInfo.ownerId
            ownerName = roomInfo.ownerName
            ownerAvatarUrl = roomInfo.ownerAvatarUrl
        }
    }

    private fun initAudienceCountView() {
        layoutAudienceCount.setOnClickListener { showAudienceListPanelView() }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initAudienceAvatarView() {
        recycleAudienceList.layoutManager =
            LinearLayoutManager(
                context,
                LinearLayoutManager.HORIZONTAL,
                false
            )
        adapter =
            AudienceListIconAdapter(
                context,
                audienceStore?.liveAudienceState
                ?: LiveAudienceStore.create(liveID = roomId).liveAudienceState
            )
        recycleAudienceList.adapter = adapter
        recycleAudienceList.setOnTouchListener(object : OnTouchListener {
            private val point = Point()
            private var scroll = false

            override fun onTouch(v: View, event: MotionEvent): Boolean {
                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        point.set(event.x.toInt(), event.y.toInt())
                        scroll = false
                    }

                    MotionEvent.ACTION_MOVE -> {
                        if (kotlin.math.abs(event.x - point.x) > 10 && !scroll) {
                            scroll = true
                        }
                    }

                    MotionEvent.ACTION_UP -> {
                        if (!scroll) {
                            showAudienceListPanelView()
                        }
                    }
                }
                return false
            }
        })
    }

    private fun showAudienceListPanelView() {
        audienceStore?.fetchAudienceList(null)
        if (audienceListPopupDialog == null) {
            audienceListPopupDialog = AudienceListPopupDialog(
                context, audienceStore?.liveAudienceState
                         ?: LiveAudienceStore.create(liveID = roomId).liveAudienceState
            )
            audienceListPopupDialog?.setOnUserItemClickListener(onUserItemClickListener)
        }
        audienceListPopupDialog?.show()
    }

    private fun onAudienceListChange(userInfoList: List<LiveUserInfo>) {
        val params = recycleAudienceList.layoutParams as RelativeLayout.LayoutParams
        if (userInfoList.size <= LIVEKIT_AUDIENCE_LIST_MAX_SHOW_AVATAR_COUNT) {
            params.width = RelativeLayout.LayoutParams.WRAP_CONTENT
        } else {
            params.width = ScreenUtil.dip2px(56F)
        }
        adapter?.updateData()
        updateAudienceCount()
    }

    private fun updateAudienceCount() {
        val userCount = audienceStore?.liveAudienceState?.audienceCount?.value ?: 0
        val listSize = audienceStore?.liveAudienceState?.audienceList?.value?.size ?: 0
        logger.info("updateAudienceCount userCount:$userCount listSize:$listSize")
        if (userCount >= LIVEKIT_AUDIENCE_LIST_ROOM_MAX_SHOW_USER_COUNT) {
            setUserCount(userCount)
        } else {
            setUserCount(listSize)
        }
    }

    @SuppressLint("SetTextI18n")
    private fun setUserCount(count: Int) {
        val text = formatUserCount(count)
        textAudienceCount.text = text
    }

    private fun formatUserCount(count: Int): String {
        var text = count.toString()
        if (count >= 10000) {
            val isChinese = Locale.CHINESE.language == Locale.getDefault().language
            if (isChinese) {
                text = String.format(Locale.getDefault(), "%.1f", count / 10000.0f)
                if (text.endsWith(".0")) {
                    text = text.replace(".0", "")
                }
                text += "万"
            } else {
                text = String.format(Locale.getDefault(), "%dk", count / 1000)
            }
        }
        return text
    }

    private fun reportData(roomId: String) {
        val isVoiceRoom = !TextUtils.isEmpty(roomId) && roomId.startsWith("voice_")
        if (isVoiceRoom) {
            reportEventData(LIVEKIT_METRICS_PANEL_SHOW_VOICE_ROOM_AUDIENCE_LIST)
        } else {
            reportEventData(LIVEKIT_METRICS_PANEL_SHOW_LIVE_ROOM_AUDIENCE_LIST)
        }
    }

    interface OnUserItemClickListener {
        fun onUserItemClick(userInfo: TUIRoomDefine.UserInfo)
    }
}